### PR TITLE
OwnerData contract supporting Yoko Ono artwork

### PR DESCRIPTION
**Pull request by tienngovan**
Story link: 
https://github.com/bitmark-inc/feral-file/issues/1953

### CODE REVIEW

File: contracts/OwnerData.sol [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+import "@openzeppelin/contracts/utils/Context.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+contract FF {
+    function ownerOf(uint256 tokenId) public view returns (address) {}
+}
+
+string constant SIGNED_MESSAGE = "Feral File is requesting authorization to write your sound piece to contract";
```
Comment (anhnguyenbitmark) [inline contracts/OwnerData.sol]:
```suggestion
string constant SIGNED_MESSAGE = "Authorize to write your data to the contract";
```

File: contracts/OwnerData.sol [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    struct Data {
+        address owner;
+        bytes dataHash;
+        string metadata;
+    }
+
+    mapping(address => mapping(uint256 => Data[])) private _tokenData;
+
+    function add(address contractAddress, uint256 tokenID, Data calldata data) external {
+        require(_isOwner(contractAddress, tokenID, _msgSender()), "OwnerData: caller is not the owner");
```
Comment (anhnguyenbitmark) [inline contracts/OwnerData.sol]: let's standardize this following ERC-6093

```solidity
// Custom errors for ERC-6093 compliance
error NotOwner(address caller, address contractAddress, uint256 tokenID);
error OwnerMismatch(address dataOwner, address caller);
```

```solidity
function add(address contractAddress, uint256 tokenID, Data calldata data) external {
    if (!_isOwner(contractAddress, tokenID, _msgSender())) {
        revert NotOwner(_msgSender(), contractAddress, tokenID);
    }
    if (data.owner != _msgSender()) {
        revert OwnerMismatch(data.owner, _msgSender());
    }
    _updateData(contractAddress, tokenID, data);
}
```

File: contracts/OwnerData.sol [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                datas[i] = datas[datas.length - 1];
+                datas.pop();
+                emit DataRemoved(contractAddress, tokenID);
+                return;
+            }
+        }
+        revert("OwnerData: data not found");
+    }
+
+    function _verifySignature(bytes memory signature) private view returns (address) {
```
Comment (anhnguyenbitmark) [inline contracts/OwnerData.sol]: There is a potential replay attack (?!)
Should add block number and verify it against a certain threshold (i.e. current block - signed block <= 100).

File: contracts/OwnerData.sol [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        bytes memory signature,
+        Data calldata data
+    ) external {
+        address signer = _verifySignature(signature);
+        require(_isOwner(contractAddress, tokenID, signer), "OwnerData: signer is not the owner");
+        require(data.owner == signer, "OwnerData: data owner mismatch");
+        _updateData(contractAddress, tokenID, data);
+    }
+
+    function remove(address contractAddress, uint256 tokenID) external {
```
Comment (anhnguyenbitmark) [inline contracts/OwnerData.sol]: Let's remove this `remove` function, as the new requirement is, that each owner can only record once on their chain, and it stays permanently (without being updated).

**Comment by anhnguyenbitmark** [comment]
Please add a batch processing ability for updating records for the first owner, so we can probably batch the request later if needed.

**Comment by anhnguyenbitmark** [comment]
Please change the base branch to master.

File: contracts/OwnerData.sol [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        Data[] storage data = _tokenData[contractAddress_][tokenID_];
+        for (uint256 i = 0; i < indexes_.length; i++) {
+            if (indexes_[i] >= data.length) {
+                revert IndexOutOfBounds();
+            }
+            if (indexes_[i] != data.length - 1) {
+                data[indexes_[i]] = data[data.length - 1];
+            }
+            data.pop();
+        }
```
Comment (anhnguyenbitmark) [inline contracts/OwnerData.sol:180]: Please add an event for this.

**Comment by anhnguyenbitmark** [comment]
For deleting, let's just replace the records with empty data. Then we won't need to sort.

File: contracts/OwnerData.sol [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+            }
+        }
+        Data[] memory result = new Data[](count);
+        for (uint256 i = 0; i < count; i++) {
+            result[i] = temp[i];
+        }
+        return result;
+    }
+
+    function remove(address contractAddress_, uint256 tokenID_, uint256[] calldata indexes_) external {
```
Comment (anhnguyenbitmark) [inline contracts/OwnerData.sol]: Please check the permissions to delete here.

File: contracts/OwnerData.sol [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+/**
+ * @title OwnerData
+ * @dev Manages addition data for ERC721 or ERC1155 token.
+ */
+contract OwnerData is Context, Ownable {
+    string private constant SIGNED_MESSAGE = "Authorize to write your data to the contract";
+    address private immutable _signer;
+    address private immutable _costReceiver;
+    uint256 public cost;
```
Comment (anhnguyenbitmark) [inline contracts/OwnerData.sol]:
```suggestion
    uint256 public serviceFee;
```

File: contracts/OwnerData.sol [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    mapping(address => mapping(uint256 => Data[])) private _tokenData;
+    // contractAddress => tokenID => owner => bool
+    mapping(address => mapping(uint256 => mapping(address => bool))) private _tokenDataOwner;
+    // contractAddress => tokenID => bool
+    mapping(address => mapping(uint256 => bool)) private _publicTokens;
+
+    event DataAdded(address indexed contractAddress, uint256 indexed tokenID, Data data);
+
+    error TrusteeIsZeroAddress();
+    error CostReceiverIsZeroAddress();
```
Comment (anhnguyenbitmark) [inline contracts/OwnerData.sol]:
```suggestion
    error EmptyServiceFeeReceiver();
```

File: contracts/OwnerData.sol [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    error PaymentRequiredForPublicToken();
+    error IndexOutOfBounds();
+    error InvalidParameters();
+    error OwnerAndSenderMismatch();
+    error SenderIsNotTheOwner();
+    error OwnerDataAlreadyAdded();
+    error InvalidSignature();
+
+
+    constructor(address signer_, address costReceiver_, uint256 cost_) {
```
Comment (anhnguyenbitmark) [inline contracts/OwnerData.sol]:
```suggestion
    constructor(address signer_, address serviceFeeReceiver_, uint256 serviceFee_) {
```

File: contracts/OwnerData.sol [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+    function signedAdd(address contractAddress_, uint256 tokenID_, Data calldata data_, Signature calldata signature_) external {
+        _validateSignature(signature_);
+        address account = data_.owner;
+        if (!_publicTokens[contractAddress_][tokenID_]) {
+            account = _recoverOwnerSignature(signature_.ownerSign);
+        }
+        _addData(account, contractAddress_, tokenID_, data_);
+    }
+
+    function _addData(address sender_, address contractAddress_, uint256 tokenID_, Data memory data_) private {
```
Comment (lpopo0856) [inline contracts/OwnerData.sol]: Add `bool isPublicToken_` so we don't have to read from storage again.

Comment (anhnguyenbitmark) [reply contracts/OwnerData.sol]: But we still need to read and verify, no?
Or we can set it immutable.

**Review by anhnguyenbitmark — APPROVED** [review]

**Pull request closed without merging**